contracts/PermissionedExchange.sol

```diff
@@ -0,0 +1,166 @@
+        exchangeController[_controller] = _isController;
+    }
+
+    /**
+     * @dev allow controllers to add the trade quota to traders on specific token
+     */
+    function addQuota(
+        address _token,
+        address _account,
+        uint256 _amount
+    ) external {
+        require(exchangeController[msg.sender] == true, 'Not controller');
+        tradeQuota[_token][_account] += _amount;
+    }
+
+    /**
+     * @dev only onwer have the permission to send the order for now,
+     * traders can do exchanges on onwer sent order
+     */
+    function sendOrder(
+        address _tokenGive,
+        address _tokenGet,
+        uint256 _amountGive,
+        uint256 _amountGet,
+        uint256 _expireDate
+    ) public onlyOwner {
+        require(_expireDate > block.timestamp, 'invalid expireDate');
+        require(_amountGive > 0 && _amountGet > 0, 'invalid amount');
+        IERC20(_tokenGive).safeTransferFrom(msg.sender, address(this), _amountGive);
+        orders[nextOrderId] = ExchangeOrder(
+            _tokenGive,
+            _tokenGet,
+            _amountGive,
+            _amountGet,
+            msg.sender,
+            _expireDate,
+            _amountGive
+        );
+        emit ExchangeOrderSent(nextOrderId, msg.sender, _tokenGive, _tokenGet, _amountGive, _amountGet, _expireDate);
+        nextOrderId += 1;
+    }
+
+    /**
+     * @dev traders do exchange on traders order, but need to trade under the trade quota.
+     */
+    function trade(uint256 _orderId, uint256 _amount) public {
+        ExchangeOrder storage order = orders[_orderId];
+        if (order.tokenGet == settings.getSQToken()) {
+            require(tradeQuota[order.tokenGet][msg.sender] >= _amount, 'tradeQuota reached');
+        }
+        require(order.expireDate > block.timestamp, 'order invalid');
+        uint256 amount = (order.amountGive * _amount) / order.amountGet;
+        require(amount <= order.amountGiveLeft, 'trade amount exceed order balance');
+        order.amountGiveLeft -= amount;
+        if (order.tokenGet == settings.getSQToken()) {
+            tradeQuota[order.tokenGet][msg.sender] -= amount;
+        }
+        IERC20(order.tokenGet).safeTransferFrom(msg.sender, order.sender, _amount);
+        IERC20(order.tokenGive).safeTransfer(msg.sender, amount);
+        emit Trade(_orderId, order.tokenGet, _amount, order.tokenGive, amount);
+    }
+
+    /**
+     * @dev everyone allowed to call settleExpiredOrder to settled expired order
+     * this will return left given token back to order sender.
+     */
+    function settleExpiredOrder(uint256 _orderId) public {
+        ExchangeOrder memory order = orders[_orderId];
+        require(order.expireDate != 0, 'order not exist');
+        require(order.expireDate < block.timestamp, 'order not expired');
+        if (order.amountGiveLeft != 0) {
+            IERC20(order.tokenGive).safeTransfer(order.sender, order.amountGiveLeft);
```

Comment:
after transfer, need set `amountGiveLeft = 0`, because if call this again, it will send again. same as below `cancalOrder`.